### PR TITLE
Add layer map search benchmark

### PR DESCRIPTION
**Pull request by bojanserafimov**
Add a layer map benchmark based on real metadata.

A few other changes in this PR:
- Move the data for the other test into a file so we don't crowd the .rs file with data
- Add a new query pattern. Our existing benchmarks all query the same key, which will probably randomly favor one solution over another, based on how favorable that key happens to be in the data structure. My new query pattern queries more keys, somewhat uniformly covering the entire search space, which is a workload similar to what we do when deciding whether we need to create a new image layer

### CODE REVIEW

File: pageserver/benches/bench_layer_map.rs [hunk 1]
```diff
@@ -203,5664 +213,46 @@ fn large_layer_map(c: &mut Criterion) {
...
+    let queries: Vec<(Key, Lsn)> = layer_map
+        .iter_historic_layers()
+        .filter_map(|l| {
+            if l.is_incremental() {
+                None
+            } else {
+                let kr = l.get_key_range();
+                let lr = l.get_lsn_range();
+
+                let mid_key = kr.start; // TODO maybe actually find the midpoint
```
Comment (knizhnik) [inline pageserver/benches/bench_layer_map.rs:238]: Just `kr.start/kr.end` will not work and find median point seems too be too expensive...
But right now name `mid_key` is confusing

File: pageserver/benches/bench_layer_map.rs [hunk 2]
```diff
@@ -178,12 +183,17 @@ fn build_layer_map() -> LayerMap {
 }
 
 fn large_layer_map(c: &mut Criterion) {
-    let layer_map = build_layer_map();
+    // A list of layer filenames, extracted from our performance test environment, from
+    // a project where we have run pgbench many timmes. The pgbench database was initialized
+    // between each test run.
+    let layer_map = build_layer_map(PathBuf::from("benches/odd-brook-layernames.txt"));
 
     c.bench_function("search", |b| {
```
Comment (knizhnik) [inline pageserver/benches/bench_layer_map.rs:191]: Why do we need this test for hardcoded key lookup if below  we perform search for real keys?

**Review by knizhnik — APPROVED** [review]

**Comment by knizhnik** [comment]
> One idea I had when adding tests around walredo was to use zstd to compress the the 8192 byte expected page to 263 bytes or so. Similarly here, zstd would compress these to 540kB => 28kB and 2.4MB => 63kB. Adding zstd as bench only dependency is however tricky as dev-dependencies cannot be optional last I checked, however bench targets can require a special feature which could enable the depedendency.

Compression can increase speed. Or can reduce it. It depends on ratio of network throughput/CPU speed. It will be interesting to make such experiment.

If compression can improve network speed, then there are several other places where compression can help.
we are sending WA between compute and safekeeper, safekeeper and pageserver, safekeeper and S3.
As far as WAL record size is small enough we should either use streaming compression, either compress some larger chunks (i.e. wal redo request). Concerning stream compression, there is PR for libpq compression support on  commitfest. I have originally written it, later it was rewritten by one guy from Yandex. Neither I neither him could push this patch to the community.
Now Borodin from Yandex tries to resurrect this patch. It can be used for compute<->safekeeper and safeeper<->pageserver communication when libpq protocol is used. But communication between pageserver and walredo  is not using libpq.

File: pageserver/benches/bench_layer_map.rs (hunk 1 above)
Comment (bojanserafimov) [reply pageserver/benches/bench_layer_map.rs:238]: Yes, I plan to fix this TODO before merging

**Comment by bojanserafimov** [comment]
Maybe I'll leave test file compression for the future. 2MB is a big file, but the repo limit is 5GB. Let me know if there are objections.

**Comment by hlinnaka** [comment]
> Maybe I'll leave test file compression for the future. 2MB is a big file, but the repo limit is 5GB. Let me know if there are objections.

No objections from me.

If you can write a piece of code that generates a similar set of layers, instead of including the list directly, that would be better. But don't sweat over it.

**Pull request merged**